---
title: "After emptyDrops (sample)"
output: html_document
author: "Kevin Rue"
date: "`r Sys.Date()`"
---

```{r libraries, include=FALSE}
library(BiocSingular)
library(BiocParallel)
library(bluster)
library(dplyr)
library(htmltools)
library(rmarkdown)
library(rtracklayer)
library(scales)
library(scater)
library(scran)
library(SingleCellExperiment)
library(tidyverse)
```

```{r constants, include=FALSE}
umi_breaks <- 10^(-1:10)
umi_minor_breaks <- rep(c(1, 2, 5), 12) * (10^rep(-1:10, each = 3))
barcode_breaks <- 10^(-1:10)
barcode_minor_breaks <- 10^rep(-1:10)
```

```{r snakemake, include=FALSE}
if (interactive()) {
  sce_rds <- "../results/emptyDrops/sce/WPPm048hrs_rep1.rds"
  mt_tsv <- "../config/mitochondrial_genes.tsv"
  gtf <- "../resources/genome/genome.gtf.gz"
  n_top_hvgs <- 500
  n_dimred_umap <- 10
  notebook_threads <- 8
} else {
  sce_rds <- snakemake@input[["sce"]]
  mt_tsv <- snakemake@input[["mt"]]
  gtf <- snakemake@input[["gtf"]]
  n_top_hvgs <- 500
  n_dimred_umap <- snakemake@params[["n_dimred_umap"]]
  notebook_threads <- snakemake@threads
}
```

## Run parameters

```{r parameters, echo=FALSE}
format_bold <- function(x) sprintf("**%s**", x)
sample_name <- fs::path_ext_remove(basename(sce_rds))
df <- matrix(
  data = c(
    format_bold("Sample name"), sample_name,
    format_bold("Input file"), sce_rds,
    format_bold("threads"), snakemake@threads
  ),
  ncol = 2, byrow = TRUE
)
knitr::kable(x = df)
```

## Load SCE object

```{r load_fry, echo=FALSE}
sce <- readRDS(sce_rds)
sce
```

## Load mitochondrial gene list

```{r, echo=FALSE}
mito_gene_ids <- read_tsv(mt_tsv, show_col_types = FALSE)[["gene_id"]]
str(mito_gene_ids)
```

## Load gene metadata

```{r}
gtf_gene_data <- import.gff(gtf, feature.type = "gene") %>% as_tibble()
```

## Barcode statistics

```{r compute_statistics, include=FALSE}
barcode_stats <- tibble(
  umi_sum = colSums(assay(sce, "counts")),
  genes = colSums(assay(sce, "counts") > 0),
  mt_pct = colSums(assay(sce, "counts")[mito_gene_ids, ]) / umi_sum
) %>%
  filter(umi_sum > 0) %>%
  arrange(desc(umi_sum)) %>%
  mutate(rank_umi = row_number()) %>%
  arrange(desc(genes)) %>%
  mutate(rank_genes = row_number())
barcode_stats
```

### Range of UMI per barcode

```{r range_umi_sum, echo=FALSE}
range(barcode_stats$umi_sum)
```

### Range of mitochondrial contents

```{r range_mt_pct, echo=FALSE}
range(barcode_stats$mt_pct)
```

### Barcode-Rank Plot

```{r barcode_rank_plot, echo=FALSE}
top_stats <- barcode_stats %>%
  slice_max(umi_sum, n = 1, with_ties = FALSE)
bottom_stats <- barcode_stats %>%
  slice_min(umi_sum, n = 1, with_ties = FALSE)
ggplot() +
  geom_line(
    aes(
      x = rank_umi,
      y = umi_sum
    ),
    data = barcode_stats
  ) +
  geom_label(
    aes(
      x = rank_umi,
      y = umi_sum,
      label = format(umi_sum, big.mark = ",")
    ),
    data = top_stats,
    alpha = 0.5,
    hjust = 0
  ) +
  geom_label(
    aes(
      x = rank_umi,
      y = umi_sum,
      label = format(umi_sum, big.mark = ",")
    ),
    data = bottom_stats,
    alpha = 0.5,
    hjust = 1
  ) +
  scale_x_log10(breaks = barcode_breaks, minor_breaks = barcode_minor_breaks, labels = label_number(scale_cut = cut_short_scale())) +
  scale_y_log10(
    limits = c(1, NA),
    breaks = umi_breaks,
    minor_breaks = umi_minor_breaks,
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  theme_bw() +
  labs(
    x = "Barcodes",
    y = "UMI Counts",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

### Features-Rank Plot

```{r feature_rank_plot, echo=FALSE}
top_stats <- barcode_stats %>%
  slice_max(genes, n = 1, with_ties = FALSE)
bottom_stats <- barcode_stats %>%
  slice_min(genes, n = 1, with_ties = FALSE)
ggplot() +
  geom_line(aes(x = rank_genes, y = genes), barcode_stats) +
  geom_label(
    aes(
      x = rank_genes,
      y = genes,
      label = format(genes, big.mark = ",")
    ),
    data = top_stats,
    alpha = 0.5,
    hjust = 0
  ) +
  geom_label(
    aes(
      x = rank_genes,
      y = genes,
      label = format(genes, big.mark = ",")
    ),
    data = bottom_stats,
    alpha = 0.5,
    hjust = 1
  ) +
  scale_x_log10(
    breaks = barcode_breaks,
    minor_breaks = barcode_minor_breaks,
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  scale_y_continuous(
    limits = c(0, NA),
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  theme_bw() +
  labs(
    x = "Barcodes",
    y = "Genes detected",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

## Mitochondrial content

```{r mito_histogram, echo=FALSE}
barcode_stats %>%
  ggplot(aes(x = mt_pct)) +
  geom_histogram(binwidth = 0.01, colour = "black", fill = "grey", linewidth = 0.1) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, NA)) +
  scale_x_continuous(labels = label_percent()) +
  scale_y_continuous(labels = label_comma()) +
  theme_bw() +
  labs(
    y = "Barcodes",
    x = "Proportion of mitochondrial counts",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

## Log-normalise

Necessary prior to HVGs.

```{r, echo=FALSE}
sce <- logNormCounts(
  x = sce,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

## HVGs

Recommended for PCA.

- Number of HVGs selected: `r n_top_hvgs`

```{r, echo=FALSE}
dec <- modelGeneVar(
  x = sce,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
```

```{r, echo=FALSE, fig.height=6, fig.width=8}
dec %>% 
  as_tibble() %>% 
  arrange(desc(bio)) %>% 
  mutate(
    rank = row_number()
  ) %>% 
  ggplot() +
  geom_line(aes(rank, bio)) +
  geom_vline(xintercept = n_top_hvgs, colour = "red", linetype = "dashed") +
  scale_x_log10(breaks = barcode_breaks, minor_breaks = umi_minor_breaks, labels = label_number(scale_cut = cut_short_scale())) +
  labs(
    x = "Rank (genes)",
    y = "Biological variance",
    subtitle = sample_name
  ) +
  theme_bw()
```

```{r, include=FALSE}
hvgs <- getTopHVGs(dec, n = n_top_hvgs)
length(hvgs)
```

```{r, echo=FALSE}
dec[["selected"]] <- factor(rownames(dec) %in% hvgs, c(TRUE, FALSE))
ggplot(dec) +
  geom_point(aes(mean, total, colour = selected), size = 0.1) +
  geom_point(aes(mean, tech), color = "black", size = 0.1) +
  scale_colour_manual(values = c("TRUE" = "deepskyblue", "FALSE" = "coral2")) +
  theme_bw()
```

## PCA

```{r, include=FALSE}
sce <- fixedPCA(
  x = sce,
  rank = 50,
  subset.row = hvgs,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

```{r, echo=FALSE, fig.height=6, fig.width=10}
plot_data <- tibble(
  percentVar = attr(reducedDim(sce, "PCA"), "percentVar")
) %>%
  mutate(
    PC = factor(row_number())
  )
ggplot(plot_data, aes(PC, percentVar)) +
  geom_col(color = "black", fill = "grey", linewidth = 0.1) +
  scale_y_continuous(limits = c(0, NA)) +
  theme_bw()
```

## UMAP

- Input dimensionality reduction: `"PCA"`
- Input dimensions: `r n_dimred_umap`

```{r, include=FALSE}
set.seed(1010)
sce <- runUMAP(
  x = sce,
  dimred = "PCA",
  n_dimred = n_dimred_umap,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

```{r, echo=FALSE, fig.height=6, fig.width=8}
reducedDim(sce, "UMAP") %>%
  as_tibble() %>%
  ggplot(aes(UMAP1, UMAP2)) +
  geom_point(size = 0.01) +
  theme_bw()
```

## Clustering

```{r cluster_barcodes, echo=FALSE}
set.seed(1010)
colData(sce)[["cluster"]] <- clusterRows(
  x = reducedDim(sce, "PCA")[, 1:50],
  BLUSPARAM = TwoStepParam(
    first = KmeansParam(
      centers = 1000,
      iter.max = 100
    ),
    second = NNGraphParam(
      shared = TRUE,
      k = 5
    )
  ),
  full = FALSE
)
table(colData(sce)[["cluster"]])
```

```{r cluster_barcodes_umap, echo=FALSE, fig.height=6, fig.width=8}
all_point_data <- reducedDim(sce, "UMAP") %>% 
  as_tibble() %>% 
  bind_cols(colData(sce) %>% as_tibble())
cluster_label_data <- reducedDim(sce, "UMAP") %>% 
  as_tibble() %>% 
  bind_cols(colData(sce) %>% as_tibble()) %>% 
  select(cluster, UMAP1, UMAP2) %>% 
  group_by(cluster) %>% 
  summarise(across(.cols = c(UMAP1, UMAP2), .fns = mean))
ggplot() +
  rasterise(geom_point(
    mapping = aes(x = UMAP1, y = UMAP2, colour = cluster),
    data = all_point_data,
    size = 0.001
  ), dpi = 300) +
  geom_text(mapping = aes(x = UMAP1, y = UMAP2, label = cluster), data = cluster_label_data) +
  theme_bw() +
  guides(
    colour = guide_legend(override.aes = list(size = 3))
  )
```

## Custom markers

```{r, echo=FALSE}
markers_list <- list(
  "Neurons" = c(
    "fne" = "FBgn0086675"
  ),
  "Cholinergic" = c(
    "ChAT" = "FBgn0000303",
    "VAChT" = "FBgn0270928"
  ),
  "Glutamatergic" = c(
    "VGlut" = "FBgn0031424",
    "Gs2" = "FBgn0001145"
  ),
  "GABAergic" = c(
    "Gad1" = "FBgn0004516",
    "VGAT" = "FBgn0033911"
  ),
  "Dopaminergic" = c(
    "ple" = "FBgn0005626",
    "Ddc" = "FBgn0000422"
  ),
  "Serotonergic" = c(
    "SerT" = "FBgn0010414",
    "Trh" = "FBgn0035187"
  ),
  "Octopaminergic / Tyraminergic" = c(
    "Tbh" = "FBgn0010329",
    "Tdc2" = "FBgn0050446"
  ),
  "Male" = c(
    "fru" = "FBgn0004652",
    "lncRNA:roX1" = "FBgn0019661",
    "lncRNA:roX2" = "FBgn0019660"
  ),
  "Female" = c(
    "Sxl" = "FBgn0264270"
  ),
  "Early/Late" = c(
    "Imp" = "FBgn0285926",
    "dati" = "FBgn0262636",
    "pdm3" = "FBgn0261588",
    "br" = "FBgn0283451"
  )
)
```

```{r, echo=FALSE, fig.height=5, fig.width=12}
for (cell_type in names(markers_list)) {
  for (cell_marker_id in markers_list[[cell_type]]){
    cell_marker_name <- gtf_gene_data %>%
      filter(gene_id == cell_marker_id) %>%
      pull(gene_name)
    stopifnot(cell_marker_id %in% rownames(sce))
    stopifnot(cell_marker_name %in% names(markers_list[[cell_type]]))
    plot_data <- reducedDim(sce, "UMAP") %>% 
      as_tibble() %>% 
      mutate(
        logcounts = as.numeric(assay(sce, "logcounts")[cell_marker_id, ])
      )
    gg_scatter <- ggplot(plot_data) +
      geom_point(aes(UMAP1, UMAP2, colour = logcounts), size = 0.01) +
      scale_colour_viridis_c() +
      labs(
        title = paste0(cell_marker_id, " | ", cell_marker_name, " (", cell_type, ")"),
        subtitle = sample_name
      ) +
      theme_bw()
    nonzero_q01 <- plot_data %>% filter(logcounts > 0) %>% pull(logcounts) %>% quantile(0.001)
    gg_histogram <- ggplot(plot_data) +
      geom_histogram(aes(x = logcounts), colour = "black", fill = "grey", binwidth = 0.1) +
      geom_vline(xintercept = nonzero_q01) +
      labs(
        title = paste0(cell_marker_id, " | ", cell_marker_name, " (", cell_type, ")"),
        subtitle = sample_name
      ) +
      theme_bw()
    print(cowplot::plot_grid(gg_scatter, gg_histogram))
  }
}
```

## Cross-tabulate cell counts

### Cluster vs Sex (predicted)

Male predicted as expression of *roX2* / *FBgn0019660* higher than the 1% quantile of log-normalised non-zero count values.

```{r}
cell_type <- "Male"
marker_gene_name <- "lncRNA:roX2"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
marker_cutoff <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  filter(logcounts > 0) %>%
  pull(logcounts) %>%
  quantile(0.01)
sce$roX2_sex <- factor(
  x = ifelse(
    test = assay(sce, "logcounts")[marker_gene_id,] >= marker_cutoff,
    yes = "male",
    no = "female"
  ),
  levels = c("female", "male")
)
table(sce$roX2_sex)
```

```{r}
cell_type <- "Male"
marker_gene_name <- "lncRNA:roX2"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
plot_data <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  bind_cols(
    colData(sce) %>% as_tibble()
  )
ggplot() +
  geom_violin(
    mapping = aes(x = cluster, y = logcounts), fill = "grey90",
    data = plot_data,
    scale = "width",
    draw_quantiles = 0.5) +
  labs(
    x ="Cluster",
    y = "Log-counts",
    title = paste0(marker_gene_id, " | ", marker_gene_name, " (", cell_type, ")"),
    subtitle = sample_name
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank()
  )
```

### Cluster vs Cholinergic (predicted)

Neuronal predicted as expression of *ChAT* / *FBgn0000303* higher than the 1% quantile of log-normalised non-zero count values.

```{r}
cell_type <- "Cholinergic"
marker_gene_name <- "ChAT"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
marker_cutoff <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  filter(logcounts > 0) %>%
  pull(logcounts) %>%
  quantile(0.01)
sce$ChAT <- factor(
  x = ifelse(
    test = assay(sce, "logcounts")[marker_gene_id,] >= marker_cutoff,
    yes = "positive",
    no = "negative"
  ),
  levels = c("positive", "negative")
)
table(sce$ChAT)
```

```{r}
cell_type <- "Cholinergic"
marker_gene_name <- "ChAT"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
plot_data <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  bind_cols(
    colData(sce) %>% as_tibble()
  )
ggplot() +
  geom_violin(
    mapping = aes(x = cluster, y = logcounts), fill = "grey90",
    data = plot_data,
    scale = "width",
    draw_quantiles = 0.5) +
  labs(
    x ="Cluster",
    y = "Log-counts",
    title = paste0(marker_gene_id, " | ", marker_gene_name, " (", cell_type, ")"),
    subtitle = sample_name
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank()
  )
```

### Cluster vs Glutamatergic (predicted)

Neuronal predicted as expression of *VGlut* / *FBgn0031424* higher than the 1% quantile of log-normalised non-zero count values.

```{r}
cell_type <- "Glutamatergic"
marker_gene_name <- "VGlut"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
marker_cutoff <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  filter(logcounts > 0) %>%
  pull(logcounts) %>%
  quantile(0.01)
sce$VGlut <- factor(
  x = ifelse(
    test = assay(sce, "logcounts")[marker_gene_id,] >= marker_cutoff,
    yes = "positive",
    no = "negative"
  ),
  levels = c("positive", "negative")
)
table(sce$VGlut)
```

```{r}
cell_type <- "Glutamatergic"
marker_gene_name <- "VGlut"
marker_gene_id <- markers_list[[cell_type]][[marker_gene_name]]
plot_data <- tibble(
  logcounts = assay(sce, "logcounts")[marker_gene_id,]
  ) %>%
  bind_cols(
    colData(sce) %>% as_tibble()
  )
ggplot() +
  geom_violin(
    mapping = aes(x = cluster, y = logcounts), fill = "grey90",
    data = plot_data,
    scale = "width",
    draw_quantiles = 0.5) +
  labs(
    x ="Cluster",
    y = "Log-counts",
    title = paste0(marker_gene_id, " | ", marker_gene_name, " (", cell_type, ")"),
    subtitle = sample_name
  ) +
  theme_bw() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank()
  )
```

## Session info

`r date()`

Bioconductor version `r BiocManager::version()`

```{r session_info, echo=FALSE}
sessionInfo()
```

```{r write_sce, include=FALSE}
saveRDS(sce, file.path("test-after-emptydrops", paste0(sample_name, ".rds")))
```
