---
title: "After emptyDrops (sample)"
output: html_document
author: "Kevin Rue"
date: "`r Sys.Date()`"
---

```{r libraries, include=FALSE}
library(BiocNeighbors)
library(BiocParallel)
library(BiocSingular)
library(bluster)
library(cowplot)
library(dplyr)
library(ggrastr)
library(htmltools)
library(rmarkdown)
library(rtracklayer)
library(scales)
library(scater)
library(scran)
library(SingleCellExperiment)
library(tidyverse)
```

```{r constants, include=FALSE}
umi_breaks <- 10^(-1:10)
umi_minor_breaks <- rep(c(1, 2, 5), 12) * (10^rep(-1:10, each = 3))
barcode_breaks <- 10^(-1:10)
barcode_minor_breaks <- 10^rep(-1:10)
```

```{r snakemake, include=FALSE}
if (interactive()) {
  sce_rds <- "../results/emptyDrops/sce/WPPp120hrs_rep2.rds"
  mt_tsv <- "../config/mitochondrial_genes.tsv"
  gtf <- "../resources/genome/genome.gtf.gz"
  n_top_hvgs <- 500
  n_dimred_umap <- 10
  notebook_threads <- 8
} else {
  sce_rds <- snakemake@input[["sce"]]
  mt_tsv <- snakemake@input[["mt"]]
  gtf <- snakemake@input[["gtf"]]
  n_top_hvgs <- 500
  n_dimred_umap <- snakemake@params[["n_pcs_umap"]]
  notebook_threads <- snakemake@threads
}
sample_name <- fs::path_ext_remove(basename(sce_rds))
```

## Run parameters

```{r parameters, echo=FALSE}
format_bold <- function(x) sprintf("**%s**", x)
df <- matrix(
  data = c(
    format_bold("Sample name"), sample_name,
    format_bold("Input file"), sce_rds,
    format_bold("threads"), notebook_threads
  ),
  ncol = 2, byrow = TRUE
)
knitr::kable(x = df)
```

## Load SCE object

```{r load_sce, echo=FALSE}
sce <- readRDS(sce_rds)
sce
```

## Load mitochondrial gene list

The character vector of gene identifiers is imported here for later use.

```{r mito_gene_ids, echo=FALSE}
mito_gene_ids <- read_tsv(mt_tsv, show_col_types = FALSE)[["gene_id"]]
str(mito_gene_ids)
```

## Load gene metadata from GTF

The metadata loaded from the GTF file is used to *replace* the `rowData` slot.

```{r gtf_gene_data, echo=FALSE}
gtf_gene_data <- import.gff(gtf, feature.type = "gene")
```

```{r add_rowdata, echo=FALSE}
new_rowdata <- gtf_gene_data %>% as_tibble() %>% as("DataFrame")
rownames(new_rowdata) <- new_rowdata[["gene_id"]]
rowData(sce) <- new_rowdata[rownames(sce), ]
colnames(rowData(sce))
```

## Barcode statistics

```{r barcode_statistics, include=FALSE}
barcode_stats <- tibble(
  umi_sum = colSums(assay(sce, "counts")),
  genes = colSums(assay(sce, "counts") > 0),
  mt_pct = colSums(assay(sce, "counts")[mito_gene_ids, ]) / umi_sum
) %>%
  filter(umi_sum > 0) %>%
  arrange(desc(umi_sum)) %>%
  mutate(rank_umi = row_number()) %>%
  arrange(desc(genes)) %>%
  mutate(rank_genes = row_number())
barcode_stats
```

### Barcode-Rank Plot

```{r barcode_rank_plot, echo=FALSE}
top_stats <- barcode_stats %>%
  slice_max(umi_sum, n = 1, with_ties = FALSE)
bottom_stats <- barcode_stats %>%
  slice_min(umi_sum, n = 1, with_ties = FALSE)
ggplot() +
  geom_line(
    aes(
      x = rank_umi,
      y = umi_sum
    ),
    data = barcode_stats
  ) +
  geom_label(
    aes(
      x = rank_umi,
      y = umi_sum,
      label = format(umi_sum, big.mark = ",")
    ),
    data = top_stats,
    alpha = 0.5,
    hjust = 0
  ) +
  geom_label(
    aes(
      x = rank_umi,
      y = umi_sum,
      label = format(umi_sum, big.mark = ",")
    ),
    data = bottom_stats,
    alpha = 0.5,
    hjust = 1
  ) +
  scale_x_log10(breaks = barcode_breaks, minor_breaks = barcode_minor_breaks, labels = label_number(scale_cut = cut_short_scale())) +
  scale_y_log10(
    limits = c(1, NA),
    breaks = umi_breaks,
    minor_breaks = umi_minor_breaks,
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  theme_bw() +
  labs(
    x = "Barcodes",
    y = "UMI Counts",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

### Features-Rank Plot

```{r feature_rank_plot, echo=FALSE}
top_stats <- barcode_stats %>%
  slice_max(genes, n = 1, with_ties = FALSE)
bottom_stats <- barcode_stats %>%
  slice_min(genes, n = 1, with_ties = FALSE)
ggplot() +
  geom_line(aes(x = rank_genes, y = genes), barcode_stats) +
  geom_label(
    aes(
      x = rank_genes,
      y = genes,
      label = format(genes, big.mark = ",")
    ),
    data = top_stats,
    alpha = 0.5,
    hjust = 0
  ) +
  geom_label(
    aes(
      x = rank_genes,
      y = genes,
      label = format(genes, big.mark = ",")
    ),
    data = bottom_stats,
    alpha = 0.5,
    hjust = 1
  ) +
  scale_x_log10(
    breaks = barcode_breaks,
    minor_breaks = barcode_minor_breaks,
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  scale_y_continuous(
    limits = c(0, NA),
    labels = label_number(scale_cut = cut_short_scale())
  ) +
  theme_bw() +
  labs(
    x = "Barcodes",
    y = "Genes detected",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

## Mitochondrial content

TODO:

* Compute 95% quantile of mitochondrial content and display as a vertical line and label,
  to guide the choice of a threshold.

```{r mito_histogram, echo=FALSE}
barcode_stats %>%
  ggplot(aes(x = mt_pct)) +
  geom_histogram(binwidth = 0.01, colour = "black", fill = "grey", linewidth = 0.1) +
  coord_cartesian(xlim = c(0, 1), ylim = c(0, NA)) +
  scale_x_continuous(labels = label_percent()) +
  scale_y_continuous(labels = label_comma()) +
  theme_bw() +
  labs(
    y = "Barcodes",
    x = "Proportion of mitochondrial counts",
    title = paste0("Sample (", sample_name, ")"),
    subtitle = paste0("Barcodes (", format(ncol(sce), big.mark = ","), ")")
  )
```

## Log-normalise

Necessary prior to HVGs.

```{r, echo=FALSE}
sce <- logNormCounts(
  x = sce,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

## HVGs

Recommended for PCA.

- Number of HVGs selected: `r n_top_hvgs`

The plots below illustrate the result of the selection strategy.

```{r, echo=FALSE}
dec <- modelGeneVar(
  x = sce,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
```

The first plot ranks genes by decreasing biological variance,
and indicates the number of HVGs selected by a vertical red dashed line.
Ideally, we want to select a sufficient number of genes with positive biological variance,
without going too far into the plateau near zero.

```{r, echo=FALSE, fig.height=6, fig.width=8}
dec %>% 
  as_tibble() %>% 
  arrange(desc(bio)) %>% 
  mutate(
    rank = row_number()
  ) %>% 
  ggplot() +
  geom_line(aes(rank, bio)) +
  geom_vline(xintercept = n_top_hvgs, colour = "red", linetype = "dashed") +
  scale_x_log10(breaks = barcode_breaks, minor_breaks = umi_minor_breaks, labels = label_number(scale_cut = cut_short_scale())) +
  labs(
    subtitle = sample_name,
    x = "Rank (genes)",
    y = "Biological variance"    
  ) +
  theme_bw()
```

```{r, include=FALSE}
hvgs <- getTopHVGs(dec, n = n_top_hvgs)
```

The second plot below highlights the selected HVGs (blue) in a scatter plot showing total variance and mean expression.
Ideally, we want to select a sufficient number of genes,
across the full range of mean expression values,
without picking genes too close to the estimated technical variance (black).

```{r, echo=FALSE}
dec[["selected"]] <- factor(rownames(dec) %in% hvgs, c(TRUE, FALSE))
ggplot(dec) +
  geom_point(aes(mean, total, colour = selected), size = 0.1) +
  geom_point(aes(mean, tech), color = "black", size = 0.1) +
  scale_colour_manual(values = c("TRUE" = "deepskyblue", "FALSE" = "coral2")) +
  labs(
    title = "Highly variable genes",
    subtitle = sample_name,
    x = "Mean expression",
    y = "Total variance"
  ) +
  theme_bw()
```

## PCA

- Number of HVGs selected: `r 50`

```{r, include=FALSE}
set.seed(1010)
sce <- fixedPCA(
  x = sce,
  rank = 50,
  subset.row = hvgs,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

The plot below displays the percentage of variance explained by each of the principal components computed.
It is intended to guide the choice of the number of principal components used in downstream analyses (e.g., clustering, UMAP).
Ideally, we want to use a sufficient number of principal components to capture most of the variance in the data set,
while reducing the dimensionality of the data set to minimise noise and reduce memory requirements for downstream analyses.

```{r, echo=FALSE, fig.height=6, fig.width=10}
plot_data <- tibble(
  percentVar = attr(reducedDim(sce, "PCA"), "percentVar")
) %>%
  mutate(
    PC = factor(row_number())
  )
ggplot(plot_data, aes(PC, percentVar)) +
  geom_col(color = "black", fill = "grey", linewidth = 0.1) +
  scale_y_continuous(limits = c(0, NA)) +
  labs(
    subtitle = sample_name,
    x = "Principal component",
    y = "Percentage variance explained"
  ) +
  theme_bw()
```

## UMAP

- Input dimensionality reduction: `"PCA"`
- Input dimensions: `r n_dimred_umap`

```{r, include=FALSE}
set.seed(1010)
sce <- runUMAP(
  x = sce,
  dimred = "PCA",
  n_dimred = n_dimred_umap,
  BPPARAM = MulticoreParam(workers = notebook_threads)
)
sce
```

```{r, echo=FALSE, fig.height=6, fig.width=8}
reducedDim(sce, "UMAP") %>%
  as_tibble() %>%
  ggplot(aes(UMAP1, UMAP2)) +
  geom_point(size = 0.01) +
  labs(
    subtitle = sample_name,
    x = "UMAP1",
    y = "UMAP2"
  ) +
  theme_bw()
```

## Clustering

```{r cluster_louvain, echo=FALSE}
for (resolution in seq(from = 0.5, to = 1.5, by = 0.1)) {
  cluster_coldata_name <- paste0("cluster_louvain_res", resolution)
  set.seed(1010)
  colData(sce)[[cluster_coldata_name]] <- clusterRows(
    x = reducedDim(sce, "PCA")[, 1:50],
    BLUSPARAM = TwoStepParam(
      first = KmeansParam(
        centers = 1000,
        iter.max = 100
      ),
      second = NNGraphParam(
        shared = TRUE,
        k = 5,
        BNPARAM = KmknnParam(
          distance = "Euclidean"
        ),
        BPPARAM = MulticoreParam(workers = notebook_threads),
        cluster.fun = "louvain",
        cluster.args = list(
          resolution = resolution
        )
      )
    ),
    full = FALSE
  )
}
```

```{r cluster_walktrap, echo=FALSE}
for (steps in 4 * 2^seq(from = 0, to = 2, by = 1)) {
  cluster_coldata_name <- paste0("cluster_walktrap_steps", steps)
  set.seed(1010)
  colData(sce)[[cluster_coldata_name]] <- clusterRows(
    x = reducedDim(sce, "PCA")[, 1:50],
    BLUSPARAM = TwoStepParam(
      first = KmeansParam(
        centers = 1000,
        iter.max = 100
      ),
      second = NNGraphParam(
        shared = TRUE,
        k = 5,
        BNPARAM = KmknnParam(
          distance = "Euclidean"
        ),
        BPPARAM = MulticoreParam(workers = notebook_threads),
        cluster.fun = "walktrap",
        cluster.args = list(
          steps = steps
        )
      )
    ),
    full = FALSE
  )
}
```

## Custom markers

```{r, echo=FALSE}
markers_list <- list(
  "Neurons" = c(
    "fne" = "FBgn0086675"
  ),
  "Cholinergic" = c(
    "ChAT" = "FBgn0000303",
    "VAChT" = "FBgn0270928"
  ),
  "Glutamatergic" = c(
    "VGlut" = "FBgn0031424",
    "Gs2" = "FBgn0001145"
  ),
  "GABAergic" = c(
    "Gad1" = "FBgn0004516",
    "VGAT" = "FBgn0033911"
  ),
  "Dopaminergic" = c(
    "ple" = "FBgn0005626",
    "Ddc" = "FBgn0000422"
  ),
  "Serotonergic" = c(
    "SerT" = "FBgn0010414",
    "Trh" = "FBgn0035187"
  ),
  "Octopaminergic / Tyraminergic" = c(
    "Tbh" = "FBgn0010329",
    "Tdc2" = "FBgn0050446"
  ),
  "Glial" = c(
    "nrv2" = "FBgn0015777"
  ),
  "Male" = c(
    "fru" = "FBgn0004652",
    "lncRNA:roX1" = "FBgn0019661",
    "lncRNA:roX2" = "FBgn0019660"
  ),
  "Female" = c(
    "Sxl" = "FBgn0264270"
  ),
  "Early/Late" = c(
    "Imp" = "FBgn0285926",
    "dati" = "FBgn0262636",
    "pdm3" = "FBgn0261588",
    "br" = "FBgn0283451"
  )
)
```

```{r markers_plots, echo=FALSE, fig.height=5, fig.width=12}
for (cell_type in names(markers_list)) {
  for (cell_marker_id in markers_list[[cell_type]]){
    cell_marker_name <- gtf_gene_data %>%
      as_tibble() %>% 
      filter(gene_id == cell_marker_id) %>%
      pull(gene_name)
    stopifnot(cell_marker_id %in% rownames(sce))
    stopifnot(cell_marker_name %in% names(markers_list[[cell_type]]))
    plot_data <- reducedDim(sce, "UMAP") %>% 
      as_tibble() %>% 
      mutate(
        logcounts = as.numeric(assay(sce, "logcounts")[cell_marker_id, ])
      ) %>% 
      bind_cols(
        colData(sce) %>% as_tibble()
      )
    gg_scatter <- ggplot(plot_data) +
      geom_point(aes(UMAP1, UMAP2, colour = logcounts), size = 0.01) +
      scale_colour_viridis_c() +
      labs(
        title = paste0(cell_marker_id, " | ", cell_marker_name, " (", cell_type, ")"),
        subtitle = sample_name
      ) +
      theme_bw()
    nonzero_q01 <- plot_data %>% filter(logcounts > 0) %>% pull(logcounts) %>% quantile(0.001)
    gg_histogram <- ggplot(plot_data) +
      geom_histogram(aes(x = logcounts), colour = "black", fill = "grey", binwidth = 0.1) +
      geom_vline(xintercept = nonzero_q01) +
      labs(
        title = paste0(cell_marker_id, " | ", cell_marker_name, " (", cell_type, ")"),
        subtitle = sample_name
      ) +
      theme_bw()
    cowplot_all <- plot_grid(gg_scatter, gg_histogram, nrow = 1)
    print(cowplot_all)
  }
}
```

## Session info

`r date()`

Bioconductor version `r BiocManager::version()`

```{r session_info, echo=FALSE}
sessionInfo()
```

```{r write_sce, include=FALSE}
saveRDS(sce, file.path("test-after-emptydrops", paste0(sample_name, ".rds")))
```
